File: backend/src/handler/user/list_user_repositories.rs
use axum::{
    extract::{Path, State},
    http::StatusCode,
};

use gitdot_api::endpoint::list_user_repositories as api;
use gitdot_core::dto::ListUserRepositoriesRequest;

use crate::{
    app::{AppError, AppResponse, AppState, AuthenticatedUser},
    dto::IntoApi,
};

#[axum::debug_handler]
pub async fn list_user_repositories(
    auth_user: Option<AuthenticatedUser>,
    State(state): State<AppState>,
    Path(user_name): Path<String>,
) -> Result<AppResponse<api::ListUserRepositoriesResponse>, AppError> {
    let viewer_id = auth_user.map(|u| u.id);
    let request = ListUserRepositoriesRequest::new(&user_name, viewer_id)?;
    state
        .user_service
        .list_repositories(request)
        .await
        .map_err(AppError::from)
        .map(|repos| AppResponse::new(StatusCode::OK, repos.into_api()))
}

File: gitdot-web/packages/api/src/endpoint/user/list-user-repositories.ts
import { z } from "zod";
import { RepositoryResource } from "../../resource";
import type { Endpoint } from "../endpoint";

export const ListUserRepositoriesRequest = z.object({});
export type ListUserRepositoriesRequest = z.infer<
  typeof ListUserRepositoriesRequest
>;

export const ListUserRepositoriesResponse = z.array(RepositoryResource);
export type ListUserRepositoriesResponse = z.infer<
  typeof ListUserRepositoriesResponse
>;

export const ListUserRepositories = {
  path: "/user/{user_name}/repositories",
  method: "GET",
  request: ListUserRepositoriesRequest,
  response: ListUserRepositoriesResponse,
} as const satisfies Endpoint;
export type ListUserRepositories = typeof ListUserRepositories;
